src/librustc/mir/transform/dataflow.rs

```diff
@@ -0,0 +1,419 @@
+            TerminatorChange::Terminator(nt) => match <R2 as Rewrite<L>>::term(&nt, l, c) {
+                TerminatorChange::None => TerminatorChange::Terminator(nt),
+                x => x
+            }
+        }
+    }
+}
+
+pub enum TerminatorChange<'tcx> {
+    /// No change
+    None,
+    /// Replace with another terminator
+    Terminator(mir::Terminator<'tcx>),
+}
+
+pub enum StatementChange<'tcx> {
+    /// No change
+    None,
+    /// Remove the statement
+    Remove,
+    /// Replace with another single statement
+    Statement(mir::Statement<'tcx>),
+    /// Replace with a list of statements
+    Statements(Vec<mir::Statement<'tcx>>),
+}
+
+impl<'tcx> StatementChange<'tcx> {
+    fn normalise(&mut self) {
+        let old = ::std::mem::replace(self, StatementChange::None);
+        *self = match old {
+            StatementChange::Statements(mut stmts) => {
+                match stmts.len() {
+                    0 => StatementChange::Remove,
+                    1 => StatementChange::Statement(stmts.pop().unwrap()),
+                    _ => StatementChange::Statements(stmts)
+                }
+            }
+            o => o
+        }
+    }
+}
+
+pub trait Transfer<'tcx, L: Lattice> {
+    type TerminatorOut;
+    /// The transfer function which given a statement and a fact produces a fact which is true
+    /// after the statement.
+    fn stmt(&mir::Statement<'tcx>, L) -> L;
+
+    /// The transfer function which given a terminator and a fact produces a fact for each
+    /// successor of the terminator.
+    ///
+    /// Corectness precondtition:
+    /// * The list of facts produced should only contain the facts for blocks which are successors
+    /// of the terminator being transfered.
+    fn term(&mir::Terminator<'tcx>, L) -> Self::TerminatorOut;
+}
+
+
+/// Facts is a mapping from basic block label (index) to the fact which is true about the first
+/// statement in the block.
+pub struct Facts<F>(Vec<F>);
+
+impl<F: Lattice> Facts<F> {
+    pub fn new() -> Facts<F> {
+        Facts(vec![])
+    }
+
+    fn put(&mut self, index: BasicBlock, value: F) {
+        let len = self.0.len();
+        self.0.extend((len...index.index()).map(|_| <F as Lattice>::bottom()));
+        self.0[index.index()] = value;
+    }
+}
+
+impl<F: Lattice> ::std::ops::Index<BasicBlock> for Facts<F> {
+    type Output = F;
+    fn index(&self, index: BasicBlock) -> &F {
+        &self.0.get(index.index()).expect("facts indexed immutably and the user is buggy!")
+    }
+}
+
+impl<F: Lattice> ::std::ops::IndexMut<BasicBlock> for Facts<F> {
+    fn index_mut(&mut self, index: BasicBlock) -> &mut F {
+        if let None = self.0.get_mut(index.index()) {
+            self.put(index, <F as Lattice>::bottom());
+        }
+        self.0.get_mut(index.index()).unwrap()
+    }
+}
+
+/// Analyse and rewrite using dataflow in the forward direction
+pub fn ar_forward<'tcx, T, P>(cfg: &CFG<'tcx>, fs: Facts<P::Lattice>, mut queue: BitVector)
```

Comment:
Could `ar_forward` just generate `queue` itself instead of taking the `queue` as an argument?